src/comparator/wrappers.py

```diff
@@ -276,6 +279,14 @@ def oneof(self) -> bool:
         """Return if the field is in oneof"""
         return self.field_pb.HasField("oneof_index")
 
+    @property
+    def proto3_optional(self) -> bool:
+        """Return if the field is proto3_optional"""
+        proto3_optional = self.field_pb.proto3_optional
+        if not self.oneof and proto3_optional:
+            raise TypeError("When proto3_optional is true, this field must belong to a oneof.")
```

Comment:
I suspect I'm misunderstanding something...
"optional" fields are permitted outside of one-ofs, I think?

Reply:
Maybe I understand it wrong for the explanations [here](https://source.corp.google.com/piper///depot/google3/net/proto2/proto/descriptor.proto;l=277) in the proto, it says `When proto3_optional is true, this field must be belong to a oneof to signal to old proto3 clients that presence is tracked for this field.` Can I get some advice here? If the proto3_optional is just a optional field for a `Field`, then yes, the checking should be moved outside the `oneof` checking. Thanks!

Reply:
My understanding is that every field marker `optional` is indeed enclosed in its own synthetic one-of. I.e. the one-of is not in the proto source.
So this code is actually probably correct, in that if `proto3_optional` is true then `self.oneof` should always also be true, as this is guaranteed by protoc.

So sorry for causing confusion!

Reply:
np! I reverted back the changes to check the oneof before `proto3_optional`. Thanks!